### ISSUE TITLE
False positive PropertyNotSetInConstructor

### ISSUE THREAD

**Issue by klimick** [vimeo/psalm#5630]
```php
<?php

trait CreatedAt {
    protected DateTimeImmutable $createdAt;

    private function onCreated(): void {
        $this->createdAt = new DateTimeImmutable();
    }
}

class Foo {
    use CreatedAt;

    public function __construct() {
    	$this->onCreated();
    }
}

class Bar extends Foo {}
```

Property `$createdAt` was initialized it `Foo` constructor. But psalm knows nothing about it in `Bar`.

https://psalm.dev/r/106a0483f7

**Comment by designermonkey** [comment]
I also have an example at: https://psalm.dev/r/9a28243243

**Comment by weirdan** [comment]
@designermonkey that's not supposed to work. You need to declare your initializer methods `final`, otherwise a descendant may override them. However Psalm should not warn when property is initialized in inherited final protected method, and it currently does: https://psalm.dev/r/20ede335c0

**Comment by designermonkey** [comment]
Interesting. Could you point me to some further reading about why thta need to be final?

Either way, the error returned isn't exactly very clear though. Is it supposed to be returning a different one in my use case?

**Comment by weirdan** [comment]
> Interesting. Could you point me to some further reading about why thta need to be final?

Well, that's fairly simple. Given class `C`:
```php
class C {
   private int $prop;

   protected function initProp(): void { $this->prop = 1; }

   public function __construct() { $this->initProp(); }

   public function getProp(): int { return $this->prop; }
}
```
it can be shown that `$this->prop` can be left uninitialized when descendant overrides `initProp()`, e.g.
```php
class D extends C {
    protected function initProp(): void {}
}

(new D)->getProp(); // will fail
```
Psalm cannot assume this won't happen, unless the method is private or final or the class itself is final.

**Comment by rulatir** [comment]
`it can be shown that $this->prop can be left uninitialized when descendant overrides initProp(), e.g.`

There is no room for "can". When a descendant overrides `initProp()` in a way that does not guarantee that `$prop` is initialized, or when it fails to invoke the implementation that does provide this guarantee (because it fails to call the parent constructor or something), then **report the problem for the descendant**. Psalm has all the information it needs in order to figure out whether the effective class, resulting from **specific, known** inheritance chain and **specific, known** sequence of trait inclusions on that chain results in a class that will properly initialize its `$prop` property. Psalm also has the ability to analyze the code of methods to determine whether the (possibly inherited) constructor of a derived class guaranteedly results in initializing a property, or whether it eventually calls code that does. There is no justification for bailing out with "sorry, dunno" here.

**Comment by weirdan** [comment]
> Psalm has all the information it needs in order to figure out whether the effective class

I wish that was true. But Psalm does not necessarily has access to all descendants - e.g. when a library ships a non-final class meant to be extended by the consumers of that library.

**Comment by rulatir** [comment]
The **crux** of this bug is the **incorrect** belief that access to descendants is required to validate ancestors. **It is not**. Psalm is not supposed to clairvoyantly report problems with a class that is perfectly correct and innocent, but its descendants might potentially do something wrong, like not calling the parent constructor or otherwise actively frustrating the base class's intention to initialize the property. When descendants **actually commit** those sins, Psalm should report problems **about those descendants** while analyzing **thir code**, at which point Psalm **will** have access to the entire inheritance chain leading to the **actual occurrence** of the problem.

**Comment by rulatir** [comment]
Lint tools' insistence on weaving morbid prophecies about an Evil Child that will be born some day are one of the big causes of the proliferation of APIs that are cripplingly locked down using `final`. Developers cut down on extensibility in order to satisfy catastrophizing lint tools.

**Comment by rulatir** [comment]
Me: Is this a constructor?
Psalm: Yes.
Me: Is the property set in it?
Psalm: Yes.
Me: Then why are you saying that PropertyNotSetInConstructor?

**Comment by SCIF** [comment]
Quite a big discussion here, but looks like even `final` is not a solution: https://psalm.dev/r/c8408355be

**Comment by SCIF** [comment]
Hmm, probably my example is actually a next bug: https://github.com/vimeo/psalm/issues/5062

**Comment by rulatir** [comment]
Seriously, this needs more urgent attention. It is absolutely unacceptable for a lint tool to force programmers to entirely throw away a huge chunk of what constitutes OO programming. This bug forces programmers to lock their libraries down with `final` all over the place, or create ultra-pedantic, bloated, resource-hungry designs with 7× more classes than necessary.

**Comment by orklah** [comment]
Hey @rulatir, feel free to take a look at it. I'd be fine with not raising this error when we can show that a parent correctly initialize the property, even if a child could theoretically override the method

**Comment by orklah** [comment]
@rulatir You seem engaged in your own little crusade against `final` that is completely out of scope here, given we stated earlier that [`final` doesn't even solve the issue](https://github.com/vimeo/psalm/issues/5630#issuecomment-820759546)

However, you managed to both massively overestimate the time it would take you to learn how to fix this issue and also massively underestimate the time it would take me (or anyone else in the project) to fix it. And you did this to justify your demand for other to fix a problem that doesn't even exists in a tool you don't use.

That's quite a performance there, pal.

This bug tracker currently has 900 open issues and many are more critical than this. So as I said, if you're concerned enough to do anything more than complain, feel free to take a look at it.

**Comment by rulatir** [comment]
My crusade is against false positives in general.

**Comment by kkmuffme** [comment]
Maybe fixed in https://github.com/vimeo/psalm/pull/11636 ?
